gneiss/util.py

```diff
@@ -0,0 +1,117 @@
+def match(x, y, intersect=False):
+    """ Sorts samples in metadata and contingency table in the same order.
+
+    Parameters
+    ----------
+    x : pd.DataFrame
+        Contingency table where samples correspond to rows and
+        features correspond to columns.
+    y: pd.DataFrame
+        Metadata table where samples correspond to rows and
+        explanatory metadata variables correspond to columns.
+    intersect : bool, optional
+        Specifies if only the intersection of samples in the
+        contingency table and the metadata table will returned.
+
+    Returns
+    -------
+    _x : pd.DataFrame
+        Filtered dataframe
+    _y : pd.DataFrame
+        Filtered dataframe
+    """
+    _x = x.sort_index()
+    _y = y.sort_index()
+    if intersect:
+        idx = set(_x.index) & set(_y.index)
+        idx = sorted(idx)
+        return _x.loc[idx], _y.loc[idx]
+    else:
+        if len(_x.index) != len(_y.index):
+            raise ValueError("`x` and `y` have incompatible sizes, "
+                             "`x` has %d rows, `y` has %d rows.  "
+                             "Consider setting `intersect=True`." %
+                             (len(_x.index), len(_y.index)))
+        return _x, _y
+
+
+def match_tips(table, tree, intersect=False):
+    """ Returns the OTU table and tree with matched tips.
+
+    Sorts the columns of the OTU table to match the tips in
+    the tree.  If the tree is multi-furcating, then the
+    tree is reduced to a bifurcating tree by randomly inserting
+    internal nodes.
+
+
+    Parameters
+    ----------
+    table : pd.DataFrame
+        Contingency table where samples correspond to rows and
+        features correspond to columns.
+    tree : skbio.TreeNode
+        Tree object where the leafs correspond to the features.
+    intersect : bool, optional
+        Specifies if only the intersection of samples in the
+        contingency table and the tree will returned.
+
+    Returns
+    -------
+    pd.DataFrame :
+        Subset of the original contingency table with the common features.
+    skbio.TreeNode :
+        Sub-tree with the common features.
+    """
+    tips = [x.name for x in tree.tips()]
+    common_tips = list(set(tips) & set(table.columns))
+
+    if intersect:
```

Comment:
That's exactly my point. Wouldn't it make more sense to have a single method that receives two (or more?) parameters, checks they are the same size and returns the sorted index? This will simplify both match* methods by removing that functionality from them ...